fix/3.x 446 facets view update remove canioncal route

andybroomfield opened the pull request:
Fix #446 

- Remove the canonical route for directory facets
- Allow anon and auth users to view directory facets

<!-- See https://docs.localgovdrupal.org/contributing/ for guidelines on contributing. -->

## What does this change?

Allow anonymous and authenticated users to view directory facets again, after the security update to facets 3.0.1 / 2.0.10

<!-- A pull request should have enough detail to be understandable far in the
future. e.g what is the problem/why is the change needed, how does it solve it
and any questions or points of discussion. Prefer copying information from a
Trello card (for example) over linking to it; the card may not always exist and
reviewers may not have access to the board. -->

## How to test

- Run the update and view a directory channel as an anonymous user.
- Try to edit and add directory facets and check functionality is still present.

<!-- Provide instructions to help others verify the change. This could take the
form of "On main, do X and witness Y. On this branch, do X and witness Z. " -->

## How can we measure success?

Directory facets are viewable again.
<!-- Do you expect errors to decrease? Do you expect user journeys to be
simplified? What can be used to prove this? A filtered view of logs or
analytics, etc? -->

## Have we considered potential risks?

Some sites might be using the directory facets canonical route.
<!-- What are the potential risks and how can they be mitigated? Does an error
require an alarm? Should user help, infosec, or legal be informed of this
change? Is private information guarded? Do we need to add anything in the
backlog? -->

## Images

Before:
<img width="1226" height="734" alt="Screenshot 2025-09-01 at 12 12 01 pm" src="https://github.com/user-attachments/assets/94b2e8f2-64f8-4d68-b7d2-86acf53ab5d6" />

After:
<img width="1220" height="841" alt="Screenshot 2025-09-01 at 12 12 37 pm" src="https://github.com/user-attachments/assets/11eecf45-ac30-49b9-932b-43dd5d6cc191" />


<!-- Usually only applicable to UI changes, what did it look like before and
what will it look like after? -->

## Accessibility

n/a

tonypaulbarker commented:
Have been attempting to test.  Both with and without the changes here I observe on a clean copy locally with demo content as user 1 empty directories channels with only the search box visible. Content appears to have installed and reference the channels with no errors apparent.

andybroomfield commented:
@tonypaulbarker I found that on fresh installs the directory search index needs to be run via search api.

tonypaulbarker commented:
@andybroomfield of course, that’s probably it. Thanks!

tonypaulbarker commented:
@andybroomfield 

The permission change appears to resolve the underlying issue tested on a fresh demo install.

There doesn't seem to be a material change for anonymous users. They appear to get a 'page not found' response instead of 'access denied' when visiting, say /admin/content/directories/facets/5. I don't reckon that makes a whole lot of difference to performance in terms of logging and crawlers.

There is a problem for editors, however and the answer to "Would anyone actually visit this page?" looks to be "yes":
Whenever adding a new facets or saving a facet after updating, say /admin/content/directories/facets/3/edit
when saving editors will see an error:

The website encountered an unexpected error. Try again later.

Drupal\Core\Entity\Exception\UndefinedLinkTemplateException: No link template 'canonical' found for the 'localgov_directories_facets' entity type in Drupal\Core\Entity\EntityBase->toUrl() (line 211 of core/lib/Drupal/Core/Entity/EntityBase.php).

Instead of the proposed changes could a role check for these routes work instead? I think we only care about accessing the facet for anonymous users?

andybroomfield commented:
@tonypaulbarker I found the culprit to this was that we generate a logger entry when a facet is saved and add a link to view the entity, which is no longer there. I've updated so we link to edit it now.

<img width="1581" height="693" alt="Screenshot 2025-09-02 at 11 53 25 am" src="https://github.com/user-attachments/assets/1d53890c-9469-46e8-977e-2b12515bb478" />

That allows facets to be edited and saved again. 
It does raise a question which I noted in the commit message, any call to 
```php
$entity->toLink()->toRenderable()
```
will fail with a directory facet with this change, because no canonical route is defined, and this is the [default returned](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21EntityInterface.php/function/EntityInterface%3A%3AtoLink/11.x).

It looks like it's quite common to remove [unwanted](https://git.drupalcode.org/project/entity_browser/-/commit/893ea180f44e6590cf9b85d2393404ab7c66b920) [canonical](https://www.drupal.org/project/address_formatter/issues/3469118) [routes](https://www.drupal.org/project/entity_share/issues/3419256) from entities if there not meant to have user facing pages.

Sounds like this needs a bit of discussion to go forward. Mean time we should recommend sites enable the view directory facets permission, even at the cost of the page being visible as otherwise they cannot upgrade to facets 2.0.10 or facets 3.0.1 and will have security warnings.

We can then decide if it is this solution or the solution I implemented for Alert banner which we forward with.

andybroomfield commented:
If we go ahead with this we should add a test?

finnlewis commented:
@tonypaulbarker testing again... once tested, if all OK, we're happy to go ahead and get this out asap so people don't lose their facets.
